### PR TITLE
Testing: Fail E2E when page displays warning notices

### PR DESCRIPTION
**Pull request by aduth**
Extracted from: #13446

This pull request seeks to add new functionality to the end-to-end tests to fail immediately fail the tests if a page loads with warning notices present.

_In Progress:_ Currently, the build fails, likely due to a race condition between page navigations and the `page.evaluate` result. See https://github.com/WordPress/gutenberg/pull/13446#issuecomment-456940674 .

**Testing instructions:**

Ensure end-to-end tests pass:

```
npm run test-e2e
```

### CODE REVIEW

**Comment by aduth** [comment]
Idea: Navigation needs to be abstracted, where the validation occurs as a pre-condition for the navigation promise being resolved. For example, this could be incorporated into `visitAdmin`.

**Comment by gziolo** [comment]
I opened #14371 where I want to provide an option to override site constansts through env variables.

**Comment by aduth** [comment]
I rebased to account for #14371, also introducing a revised approach which (per https://github.com/WordPress/gutenberg/pull/13452#issuecomment-458194694) checks for errors in the markup as part of `visitAdminPage`. This should also improve on the previous implementation which would only have caught notices which were output before the start of the page markup, when errors could in-fact occur anywhere in markup. It does so testing for the specific pattern of PHP error markup.

File: packages/e2e-test-utils/src/has-php-error.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
+/**
+ * Regular expression matching a displayed PHP error within a markup string.
+ *
+ * @link https://github.com/php/php-src/blob/598175e/main/main.c#L1257-L1297
+ *
+ * @type RegExp
+ */
+const REGEXP_PHP_ERROR = /<b>(Fatal error|Recoverable fatal error|Warning|Parse error|Notice|Strict Standards|Deprecated|Unknown error)<\/b>/;
```
Comment (gziolo) [inline packages/e2e-test-utils/src/has-php-error.js]: Should this match be more detailed? `<b>Notice</b>` or `<b>Warning</b>` might be too common and cause false positives.  We could at least add `:` after closing tag:
https://github.com/php/php-src/blob/598175e/main/main.c#L1321

**Review by gziolo — COMMENTED** [review]
It still needs `npm run docs:build` :)

File: packages/e2e-test-utils/src/has-php-error.js (hunk 1 above)
Comment (aduth) [reply packages/e2e-test-utils/src/has-php-error.js]: Yeah, I suppose it risks false positives in its current form. With the placeholder formatting and multiple variants, it's harder to get a good pattern to cover all the error outputs. I agree the colon `:` and even the subsequent double-space `  ` could be a good improvement to eliminate those false positives. I'll make the change.

Comment (gziolo) [reply packages/e2e-test-utils/src/has-php-error.js]: Thank you, I will give it a spin as code wise it look good 👍

**Comment by gziolo** [comment]
I don't know why but this is how notices and warnings are printed in my local copy of Docker for Gutenberg:
```
Notice: Undefined property: WP_Block_Type_Registry::$x in /var/www/html/wp-content/plugins/gutenberg/lib/blocks.php on line 47
--
  |  
  | Warning: Cannot modify header information - headers already sent by (output started at /var/www/html/wp-content/plugins/gutenberg/lib/blocks.php:47) in /var/www/html/wp-admin/includes/misc.php on line 1198

```

**Comment by aduth** [comment]
@gziolo Hmm, by the code of `main.c`, there's a few variations of how the errors are displayed, based on the result of `html_errors` (some PHP configuration?).

https://github.com/php/php-src/blob/598175e/main/main.c#L1331
https://github.com/php/php-src/blob/598175e/main/main.c#L1336
https://github.com/php/php-src/blob/598175e/main/main.c#L1306

I suppose one option is to make the `<b>` matching optional, though the downside to this is re-opening the possibility of false positives.

Another option which comes to mind is to direct the PHP errors to write to a log file, and consider a build as successful only if the log file doesn't exist / is empty. Not sure off-hand how easy this would be to implement.

**Comment by aduth** [comment]
`html_errors` is a runtime configuration:

https://www.php.net/manual/en/errorfunc.configuration.php#ini.html-errors

Perhaps we could normalize this behavior to ensure HTML output. A matter of calling `ini_set( "html_errors", 1 );` ?

**Comment by gziolo** [comment]
> Another option which comes to mind is to direct the PHP errors to write to a log file, and consider a build as successful only if the log file doesn't exist / is empty. Not sure off-hand how easy this would be to implement.

Given that you can use any website to run e2e tests, I don't think that using log file would work as you might not have access to read from the filesystem in the place where such log file is stored. 

This makes me wonder whether we should put it behind the flag and optimize against Travis and don't care too much about all other use cases?

**Comment by youknowriad** [comment]
Should we just land this? it seems still relevant?

**Comment by aduth** [comment]
It is relevant, and still useful. Based on earlier conversation, I think there were some challenges around particular configuration setups. We could maybe ship a semi-working implementation and improve it later.

**Comment by youknowriad** [comment]
This works for me, I'd be curious to know if the tests still pass after a rebase and we can ship

**Comment by aduth** [comment]
Rebased to bring this back up to date.

A few changes:

1. I renamed `hasPHPError` to `hasPageError`. There's no reason for it to be so specific to PHP. Page-level errors could come from a multitude of sources. The implementation itself didn't change, and it's fine to target those PHP errors internally.
2. Expanded regular expression to match both plain-text and HTML error outputs.
3. Added unit tests.

**Review by youknowriad — APPROVED** [review]
LGTM 👍

**Comment by aduth** [comment]
I ran through a final round of testing to see how it worked. It did work correctly, but the error output isn't especially helpful:

```
  ● new editor state › should be immediately saveable

    expect(received).toBe(expected) // Object.is equality

    Expected: false
    Received: true

      at _callee$ (../e2e-test-utils/build/@wordpress/e2e-test-utils/src/visit-admin-page.js:20:8)
      at tryCatch (../../node_modules/regenerator-runtime/runtime.js:45:40)
      at Generator.invoke [as _invoke] (../../node_modules/regenerator-runtime/runtime.js:271:22)
      at Generator.prototype.<computed> [as next] (../../node_modules/regenerator-runtime/runtime.js:97:21)
      at asyncGeneratorStep (../../node_modules/@babel/runtime/helpers/asyncToGenerator.js:3:24)
      at _next (../../node_modules/@babel/runtime/helpers/asyncToGenerator.js:25:9)
```

It's also not entirely clear how to customize this messaging from within Jest (see https://github.com/facebook/jest/issues/3293). I think it can probably be done by either throwing a custom error, or by using [Node's internal `assert` module](https://nodejs.org/api/assert.html). Thinking about this had me contemplating that it would probably be helpful to include the notice itself as part of the messaging output. I think it could be relatively straight-forward to change `hasPageError` to `getPageError`, where with the presence of a non-empty value we can output the message as part of the failure message.

I'll plan to put this together in the morning.

**Comment by aduth** [comment]
> Thinking about this had me contemplating that it would probably be helpful to include the notice itself as part of the messaging output. I think it could be relatively straight-forward to change `hasPageError` to `getPageError`, where with the presence of a non-empty value we can output the message as part of the failure message.

Updated in 965900a and 228aad3.

New messaging example:

```
  ● new editor state › should be immediately saveable

    Unexpected error in page content: Notice: Undefined offset: 1 in /var/www/html/wp-content/plugins/gutenberg/gutenberg.php on line 164

      at _callee$ (../e2e-test-utils/build/@wordpress/e2e-test-utils/src/visit-admin-page.js:30:9)
      at tryCatch (../../node_modules/regenerator-runtime/runtime.js:45:40)
      at Generator.invoke [as _invoke] (../../node_modules/regenerator-runtime/runtime.js:271:22)
      at Generator.prototype.<computed> [as next] (../../node_modules/regenerator-runtime/runtime.js:97:21)
      at asyncGeneratorStep (../../node_modules/@babel/runtime/helpers/asyncToGenerator.js:3:24)
      at _next (../../node_modules/@babel/runtime/helpers/asyncToGenerator.js:25:9)
```

**Pull request merged**